File: lab-test-env/memory-for-ai/scripts/check_knowledge.py
#!/usr/bin/env python3
"""
check_knowledge.py — Validates KNOWLEDGE.json for the Memory for AI lab.

Run: python scripts/check_knowledge.py
"""

import json
import sys
from pathlib import Path

KNOWLEDGE_PATH = Path(__file__).parent.parent / "KNOWLEDGE.json"
MIN_ENTRIES = 3
REQUIRED_FIELDS = ["concept", "sql_instinct_overridden", "rule", "when_to_apply", "confidence"]
VALID_CONFIDENCE = {"verified", "corrected", "self-assessed"}

# Lab-specific coverage checks: at least one entry must match each
COVERAGE_CHECKS = [
    {
        "label": "namespace isolation or multi-tenant memory",
        "test": lambda e: any(
            kw in f"{e.get('concept','')} {e.get('rule','')} {e.get('when_to_apply','')}".lower()
            for kw in ["namespace", "isolation", "tenant", "user"]
        ),
    },
    {
        "label": "vector search or semantic retrieval",
        "test": lambda e: any(
            kw in f"{e.get('concept','')} {e.get('rule','')} {e.get('when_to_apply','')}".lower()
            for kw in ["vector", "embed", "semantic", "retriev"]
        ),
    },
    {
        "label": "short-term vs long-term memory distinction",
        "test": lambda e: any(
            kw in f"{e.get('concept','')} {e.get('rule','')}".lower()
            for kw in ["short", "long", "checkpointer", "saver", "store", "persist"]
        ),
    },
]


def check_knowledge():
    results = []
    passed = True

    # 1. File exists
    if not KNOWLEDGE_PATH.exists():
        print("✗ KNOWLEDGE.json not found.")
        print("  Create KNOWLEDGE.json in the lab root after completing the lab.")
        print("\nKnowledge Check: FAIL")
        sys.exit(1)
    results.append("✓ KNOWLEDGE.json exists")

    # 2. Valid JSON
    try:
        with open(KNOWLEDGE_PATH, "r", encoding="utf-8") as f:
            entries = json.load(f)
    except json.JSONDecodeError as e:
        print(f"✗ KNOWLEDGE.json: invalid JSON — {e}")
        print("\nKnowledge Check: FAIL")
        sys.exit(1)
    results.append("✓ Valid JSON")

    # 3. Is a list
    if not isinstance(entries, list):
        print("✗ KNOWLEDGE.json must be a JSON array of knowledge entries")
        print("\nKnowledge Check: FAIL")
        sys.exit(1)
    results.append("✓ Is a JSON array")

    # 4. Minimum entry count
    if len(entries) < MIN_ENTRIES:
        results.append(
            f"✗ Entry count: {len(entries)} (need ≥ {MIN_ENTRIES}). "
            "Add a knowledge entry for each major concept you learned."
        )
        passed = False
    else:
        results.append(f"✓ Entry count: {len(entries)} (≥ {MIN_ENTRIES})")

    # 5. Required fields on every entry
    fields_ok = True
    for i, entry in enumerate(entries):
        concept_label = entry.get("concept", "unnamed")
        missing = [f for f in REQUIRED_FIELDS if not str(entry.get(f, "")).strip()]
        if missing:
            results.append(f'✗ Entry {i + 1} ("{concept_label}"): missing fields — {", ".join(missing)}')
            fields_ok = False
        conf = entry.get("confidence", "")
        if conf and conf not in VALID_CONFIDENCE:
            results.append(
                f'✗ Entry {i + 1}: confidence must be one of: '
                f'{", ".join(sorted(VALID_CONFIDENCE))} (got "{conf}")'
            )
            fields_ok = False
    if fields_ok:
        results.append("✓ All entries have required fields with valid confidence values")
    else:
        passed = False

    # 6. Lab-specific coverage
    for check in COVERAGE_CHECKS:
        covered = any(check["test"](e) for e in entries)
        if covered:
            results.append(f"✓ Coverage: {check['label']}")
        else:
            results.append(f"✗ Coverage: no entry covers {check['label']}")
            passed = False

    for r in results:
        print(r)

    print()
    if passed:
        print("Knowledge Check: PASS")
        sys.exit(0)
    else:
        print("Knowledge Check: FAIL")
        sys.exit(1)


if __name__ == "__main__":
    check_knowledge()

File: lab-test-env/builder-badge/scripts/check-knowledge.js
'use strict';

const fs = require('fs');
const path = require('path');

const KNOWLEDGE_PATH = path.join(__dirname, '..', 'KNOWLEDGE.json');
const MIN_ENTRIES = 4;
const REQUIRED_FIELDS = ['concept', 'sql_instinct_overridden', 'rule', 'when_to_apply', 'confidence'];
const VALID_CONFIDENCE = ['verified', 'corrected', 'self-assessed'];

// Lab-specific coverage: at least one entry must touch each theme
const COVERAGE_CHECKS = [
  {
    label: 'schema design (embedding or referencing)',
    test: entry => /embed|reference|schema/i.test(`${entry.concept} ${entry.rule} ${entry.when_to_apply}`)
  },
  {
    label: 'vector search or indexing',
    test: entry => /vector|index/i.test(`${entry.concept} ${entry.rule} ${entry.when_to_apply}`)
  }
];

function checkKnowledge() {
  const results = [];
  let passed = true;

  // 1. File exists
  if (!fs.existsSync(KNOWLEDGE_PATH)) {
    console.log('✗ KNOWLEDGE.json: not found');
    console.log('  Create KNOWLEDGE.json in the lab root. See README for schema.');
    console.log('\nKnowledge Check: FAIL');
    process.exit(1);
  }
  results.push('✓ KNOWLEDGE.json: exists');

  // 2. Valid JSON
  let entries;
  try {
    const raw = fs.readFileSync(KNOWLEDGE_PATH, 'utf8');
    entries = JSON.parse(raw);
  } catch (err) {
    results.push(`✗ KNOWLEDGE.json: invalid JSON — ${err.message}`);
    results.forEach(r => console.log(r));
    console.log('\nKnowledge Check: FAIL');
    process.exit(1);
  }
  results.push('✓ KNOWLEDGE.json: valid JSON');

  // 3. Is an array
  if (!Array.isArray(entries)) {
    results.push('✗ KNOWLEDGE.json: must be a JSON array of knowledge entries');
    results.forEach(r => console.log(r));
    console.log('\nKnowledge Check: FAIL');
    process.exit(1);
  }

  // 4. Minimum entry count
  if (entries.length < MIN_ENTRIES) {
    results.push(`✗ Entry count: ${entries.length} (need ≥ ${MIN_ENTRIES}). Add more knowledge entries.`);
    passed = false;
  } else {
    results.push(`✓ Entry count: ${entries.length} (≥ ${MIN_ENTRIES})`);
  }

  // 5. Required fields on each entry
  entries.forEach((entry, i) => {
    const missing = REQUIRED_FIELDS.filter(f => !entry[f] || String(entry[f]).trim() === '');
    if (missing.length > 0) {
      results.push(`✗ Entry ${i + 1} ("${entry.concept || 'unnamed'}"): missing fields — ${missing.join(', ')}`);
      passed = false;
    }
    if (entry.confidence && !VALID_CONFIDENCE.includes(entry.confidence)) {
      results.push(`✗ Entry ${i + 1}: confidence must be one of: ${VALID_CONFIDENCE.join(', ')} (got "${entry.confidence}")`);
      passed = false;
    }
  });

  if (passed) {
    results.push('✓ All entries have required fields with valid confidence values');
  }

  // 6. Lab-specific coverage
  COVERAGE_CHECKS.forEach(({ label, test }) => {
    const covered = entries.some(test);
    if (covered) {
      results.push(`✓ Coverage: ${label}`);
    } else {
      results.push(`✗ Coverage: no entry covers ${label}`);
      passed = false;
    }
  });

  results.forEach(r => console.log(r));
  console.log('');
  console.log(passed ? 'Knowledge Check: PASS' : 'Knowledge Check: FAIL');
  if (!passed) process.exit(1);
}

checkKnowledge();
